samza-core/src/main/java/org/apache/samza/runtime/AbstractApplicationRunner.java

```diff
@@ -81,8 +81,17 @@ ExecutionPlan getExecutionPlan(StreamApplication app, String runId, StreamManage
         ? ApplicationMode.BATCH : ApplicationMode.STREAM;
     cfg.put(ApplicationConfig.APP_MODE, mode.name());
 
+    // merge user-provided configuration with input/output descriptor generated configuration
+    // descriptor generated configuration has higher priority
+    Map<String, String> systemStreamConfigs = new HashMap<>();
+    graphSpec.getInputDescriptors().forEach((key, value) -> systemStreamConfigs.putAll(value.toConfig()));
+    graphSpec.getOutputDescriptors().forEach((key, value) -> systemStreamConfigs.putAll(value.toConfig()));
+    graphSpec.getSystemDescriptors().forEach(sd -> systemStreamConfigs.putAll(sd.toConfig()));
+    Map<String, String> appConfigs = new HashMap<>(cfg);
+    appConfigs.putAll(systemStreamConfigs);
+
     // create the physical execution plan
-    ExecutionPlanner planner = new ExecutionPlanner(new MapConfig(cfg), streamManager);
+    ExecutionPlanner planner = new ExecutionPlanner(new MapConfig(appConfigs), streamManager);
```

Comment:
From the SEP:
Step 3: Merge the user provided configuration (application.cfg) with the Descriptor generated configuration (System/Input/OutputDescriptor#toConfig). Descriptor generated configuration has higher precedence and overrides user provided configuration.

Step 4: Run the planning phase using this merged configuration. The planner internally creates StreamSpec instances based on this merged configuration, mutates these StreamSpecs to set additional inferred stream properties, creates/validates internal streams using them, and then merges StreamSpec generated configuration with previously merged configuration from Step 3. StreamSpec generated configuration has higher precedence.

--------

in LocalApplicationRunner#run -> AbstractApplicationRunner#getExecutionPlan (creates the plan) -> ExecutionPlanner#updateExistingPartitions (fetch the external streams partition info) which retrieve the partition counts for the streams in this system using SystemAdmins, but at this point the i noticed that Descriptor generated configs aren't present when I tried testing. This configs are needed for generating SystemAdmins.  

Please clarify the flow. Thanks!

Reply:
Good catch, thanks. Looks like the StreamManager is created using the original configs passed to the LocalApplicationRunner constructor. I'll try to make it use the configs generated by the planner.

Reply:
PS: To take this merged configuration I think we have to buildAndStartStreamManager() after we merge the configs (in ExecutionPlan) so that SystemAdmins can be initialized. 

Currently this happens in opposite order in LocalApplicationRunner
```
 @Override
  public void run(StreamApplication app) {
    StreamManager streamManager = null;
    try {
      streamManager = buildAndStartStreamManager();

      // 1. initialize and plan
      ExecutionPlan plan = getExecutionPlan(app, streamManager);

...
```

In addition to thisTest Framework should assembles all the config in the same way as done in AbstractApplicationRunner to pass overriden configs from InMemory system and stream descriptors to LocalApplicationRunner

Reply:
That's right, stream manager should ideally be built after planning.
Can you clarify what was broken for you? Any reproducible tests?

Can you clarify why test framework should reassemble configs like ApplicationRunner?

Reply:
Any Testing Integration test will be broken with following exception since there are no SystemAdmins available because the config (system -> factory) is available after the execution plan (after merger) and we instantiate  streamManager = buildAndStartStreamManager();
 before 
```
Caused by: org.apache.samza.SamzaException: Cannot get systemAdmin for system test
	at org.apache.samza.system.SystemAdmins.getSystemAdmin(SystemAdmins.java:63)
	at org.apache.samza.execution.StreamManager.getStreamPartitionCounts(StreamManager.java:91)
	at org.apache.samza.execution.ExecutionPlanner.updateExistingPartitions(ExecutionPlanner.java:167)
	at org.apache.samza.execution.ExecutionPlanner.plan(ExecutionPlanner.java:74)
	at org.apache.samza.runtime.AbstractApplicationRunner.getExecutionPlan(AbstractApplicationRunner.java:95)
	at org.apache.samza.runtime.AbstractApplicationRunner.getExecutionPlan(AbstractApplicationRunner.java:62)
	at org.apache.samza.runtime.LocalApplicationRunner.run(LocalApplicationRunner.java:166)
	... 24 more
```


Once above issue is fixed, TestFramework will still not work, lets see a sample app here

```
 final StreamApplication pageViewRepartition = (streamGraph, cfg) -> {
    KafkaSystemDescriptor ksd = new KafkaSystemDescriptor("test");
    KafkaInputDescriptor<KV<String, PageView>> isd =
        ksd.getInputDescriptor("PageView", KVSerde.of(new NoOpSerde<>(), new NoOpSerde<>()));
    MessageStream<KV<String, TestData.PageView>> inputStream = streamGraph.getInputStream(isd);
    inputStream
        .map(Values.create())
        .partitionBy(PageView::getMemberId, pv -> pv, "p1")
        .sink((m, collector, coordinator) -> {
            collector.send(new OutgoingMessageEnvelope(new SystemStream("test", "Output"),
                m.getKey(), m.getKey(),
                m));
          });
  };
```

This results in configs that assign the system defined here `test` a KafkaSystemFactory, now this can be only overriden by passing a config `jobs.jobs-id.system....factory`. 

This config value if  InMemorySystemFactory which can be only by TestRunner since the existence of InMemoryInputDescriptor, InMemoryOutputDescriptor and InMemorySystemDescriptoe is outside the user code in the TestRunner api as given below

```
 InMemorySystemDescriptor isd = new InMemorySystemDescriptor("test");
    InMemoryInputDescriptor<PageView> imid = isd.getInputDescriptor("PageView", new NoOpSerde<>());
        imid.initialize(pageviews);
    InMemoryOutputDescriptor imod = isd.getOutputDescriptor("Output", new NoOpSerde<>()).withPartitionCount(10);

    TestRunner
        .of(pageViewRepartition)
        .addInputStream(imid)
        .addOutputStream(imod)
        .addOverrideConfig("job.default.system", "test")
        .run(Duration.ofMillis(1500));
```

Hence I assemble configs for System and streams from`imid` and 'imod' above in the same precedence order as defined by your code and combine them with user defined configs and pass to LocalApplicationRunner. This configs will now get further appended by your kafka configs from code which won't take precedence!

Let me know if any concerns

Reply:
Discussed with @Sanil15 offline. We'll fix the stream manager config bug: https://github.com/apache/samza/pull/612/files.